T34559 migrate to 16 alpha

manuq opened the pull request:
Please merge this together with endlessm/kolibri-explore-plugin#520.

https://phabricator.endlessm.com/T34559

dylanmccall commented on src/kolibri_android/main_activity/activity.py:
```diff
@@ -186,13 +186,15 @@ def replace_url(self, url):
 
         self._kolibri_bus = _build_kolibri_process_bus(self)
-        app_key = self._kolibri_bus.get_app_key()
+        app_key = DeviceAppKey.get_app_key()
```
This shouldn't be necessary. [`KolibriAppProcessBus.get_app_key()`](https://github.com/endlessm/kolibri-installer-android/blob/develop/src/kolibri_android/main_activity/kolibri_bus.py#L28-L29) is our own addition ;) It's like that to keep imports from `kolibri` in one place.

dylanmccall committed:
Jenkins: Change Kolibri WHL URL to v0.16.0-alpha12

https://phabricator.endlessm.com/T34559

dylanmccall committed:
Remove duplicate code from android_whitenoise

By adding a patch to lightly refactor kolibri_whitenoise, we can change
android_whitenoise to use a subclass of DynamicWhiteNoise.

It is still necessary to copy DynamicWhiteNoise in order to subclass it,
because the same class in Kolibri itself needs to be monkey-patched with
our version. However, that copy can now be created automatically.

https://phabricator.endlessm.com/T34559

rtibbles commented on src/kolibri_android/kolibri_utils.py:
```diff
@@ -124,7 +126,7 @@ def _monkeypatch_whitenoise():
 
     logger.info("Applying DynamicWhiteNoise workarounds")
-    kolibri_whitenoise.DynamicWhiteNoise = DynamicWhiteNoise
+    kolibri_whitenoise.DynamicWhiteNoise = AndroidDynamicWhiteNoise
```
The alternative to monkey patching this is to just define your own WSGI application instead, I would think, which may feel neater? Might be a few more lines of code though.

dylanmccall replied:
Yeah, I should play with that properly :) The reason I didn't here is because Kolibri's wsgi.py has a fair bit of stuff in it. Looking at it again, it doesn't change very frequently, so not really that scary, and it would indeed make the rest of it simpler.

dylanmccall replied:
Okay, I gave it a try in #121. I agree, I like it better that way :+1:

dbnicholson commented on src/kolibri_android/android_whitenoise.py:
```diff
@@ -1,88 +1,130 @@
 from .android_utils import stat_file
 
+# We import from evil_kolibri so we can monkey-patch DynamicWhiteNoise from
+# kolibri without creating a circular dependency.
```
I don't see anything that installs an `evil_kolibri` package. Did I miss something?

dylanmccall replied:
Oops, good catch! I forgot to commit that. Fixed now in the newest version of this commit, with some new stuff added to the Makefile: copying some code from `kolibri` to `evil_kolibri`, and applying the patch I had added.

dbnicholson reviewed (approved):
I trust this does the right thing and the copying is necessary.
